feat: Replacing jsonwebtoken with fast-jwt (#184)
fastify/fastify-jwt: README.md

[![Known Vulnerabilities](https://snyk.io/test/github/fastify/fastify-jwt/badge.svg)](https://snyk.io/test/github/fastify/fastify-jwt)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat)](https://standardjs.com/)

JWT utils for Fastify, internally it uses [fast-jwt](https://github.com/nearform/fast-jwt).

**NOTE:** The plugin has been migrated from using `jsonwebtoken` to `fast-jwt`. Even though `fast-jwt` has 1:1 feature implementation with `jsonwebtoken`, some _exotic_ implementations might break. In that case please open an issue with details of your implementation.

> In addition, note that if you made use of strings describing a time span [vercel/ms](https://github.com/vercel/ms) (eg: `"1h"`, `"10 days"`, `"1 month"`...), you **must** convert all these references to **milliseconds**.

Author: This has been updated. The additional note was removed in https://github.com/fastify/fastify-jwt/pull/184/commits/3665579076b65f0cdcefe6fd0f797524fdd2e740.

[![Known Vulnerabilities](https://snyk.io/test/github/fastify/fastify-jwt/badge.svg)](https://snyk.io/test/github/fastify/fastify-jwt)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat)](https://standardjs.com/)

JWT utils for Fastify, internally it uses [fast-jwt](https://github.com/nearform/fast-jwt).

> **NOTE:** The plugin has been migrated from using `jsonwebtoken` to `fast-jwt`. Even though `fast-jwt` has 1:1 feature implementation with `jsonwebtoken`, some _exotic_ implementations might break. In that case please open an issue with details of your implementation.

Author: Good catch!

> **NOTE:** The plugin has been migrated from using `jsonwebtoken` to `fast-jwt`. Even though `fast-jwt` has 1:1 feature implementation with `jsonwebtoken`, some _exotic_ implementations might break. In that case please open an issue with details of your implementation.

Reviewer: Since we are talking about this migration, could we mention a simple upgrade list?

Mainly it concerns the fields name from `issuer` to `iss` etc..

> **NOTE:** The plugin has been migrated from using `jsonwebtoken` to `fast-jwt`. Even though `fast-jwt` has 1:1 feature implementation with `jsonwebtoken`, some _exotic_ implementations might break. In that case please open an issue with details of your implementation.

Reviewer 2: Good point.
Examples need an updated to reflect those changes

  // Global default signing method options
  sign: {
    algorithm: 'ES256',

>     isss: 'api.example.tld'

Reviewer 2: ```suggestion
    iss: 'api.example.tld'
```

Author: Whoops held my finger too long on the keyboard 😸

### `decode`

* `complete`: Return an object with the decoded header, payload, signature and input (the token part before the signature), instead of just the content of the payload. Default is `false`.
* `checkTyp`: When validating the decoded header, setting this option forces the check of the typ property against this value. Example: `checkTyp: 'JWT'`. Default is `undefined`.

### `sign`

* `key`: A string or a buffer containing the secret for `HS*` algorithms or the PEM encoded public key for `RS*`, `PS*`, `ES*` and `EdDSA` algorithms. The key can also be a function accepting a Node style callback or a function returning a promise. If provided, it will override the value of [secret](#secret-required) provided in the options.
* `algorithm`: The algorithm to use to sign the token. The default is autodetected from the key, using RS256 for RSA private keys, HS256 for plain secrets and the correspondent ES or EdDSA algorithms for EC or Ed* private keys.

> * `mutatePayload`: If the original payload must be modified in place (via Object.assign) and thus will result changed to the caller funciton.

Reviewer 2: ```suggestion
* `mutatePayload`: If `true`, the original payload will be modified in place (via `Object.assign`) by the signing function. This is useful if you need a raw reference to the payload after claims have been applied to it but before it has been encoded into a token.
```
Forgot to change this sentence to make it more clear.

> * `mutatePayload`: If the original payload must be modified in place (via Object.assign) and thus will result changed to the caller funciton.

Reviewer 2: ```suggestion
* `mutatePayload`: If `true`, the original payload will be modified in place (via `Object.assign`) by the signing function. This is useful if you need a raw reference to the payload after claims have been applied to it but before it has been encoded into a token.
```

* ... the rest of the **sign** options can be found [here](https://github.com/nearform/fast-jwt#createsigner).

### `verify`

* `key`: A string or a buffer containing the secret for `HS*` algorithms or the PEM encoded public key for `RS*`, `PS*`, `ES*` and `EdDSA` algorithms. The key can also be a function accepting a Node style callback or a function returning a promise. If provided, it will override the value of [secret](#secret-required) provided in the options.
* `algorithms`: List of strings with the names of the allowed algorithms. By default, all algorithms are accepted.
* `complete`: Return an object with the decoded header, payload, signature and input (the token part before the signature), instead of just the content of the payload. Default is `false`.

> * `cache`: A positive number specifying the size of the verified tokens cache (using LRU strategy). Setting this to `true` is equivalent to provide the size 1000. When enabled the  performance is dramatically improved. By default the cache is disabled.

Reviewer 3: I would go ahead and enable the cache by default.

Wdyt?

Reviewer 2: Unless it is enabled by default in `fast-jwt` I would not make it `true` by default in `fastify-jwt`.

But we can properly document this feature (with an example use case and maybe a link to [this NearForm article](https://www.nearform.com/blog/improve-json-web-tokens-performance-in-node-js/)) and highlight the benefits in the `README.md` and in the upcoming package changelog.

Author: I agree with @darkgl0w

* `expiresIn`: Time span (in milliseconds) after which the token expires, added as the exp claim in the payload. This will override any existing value in the claim.

* ... the rest of the **sign** options can be found [here](https://github.com/nearform/fast-jwt#createsigner).

### `verify`

* `key`: A string or a buffer containing the secret for `HS*` algorithms or the PEM encoded public key for `RS*`, `PS*`, `ES*` and `EdDSA` algorithms. The key can also be a function accepting a Node style callback or a function returning a promise. If provided, it will override the value of [secret](#secret-required) provided in the options.
* `algorithms`: List of strings with the names of the allowed algorithms. By default, all algorithms are accepted.
* `complete`: Return an object with the decoded header, payload, signature and input (the token part before the signature), instead of just the content of the payload. Default is `false`.
* `cache`: A positive number specifying the size of the verified tokens cache (using LRU strategy). Setting this to `true` is equivalent to provide the size 1000. When enabled the  performance is dramatically improved. By default the cache is disabled.

> * `cacheTTL`: The maximum time to live of a cache entry (in milliseconds). If the token has a earlier expiration or the verifier has a shorter `maxAge`, the earlier takes precedence. The default is `600000`, which is 10 minutes.

Reviewer 2: ```suggestion
* `cacheTTL`: The maximum time to live of a cache entry (in milliseconds). If the token has a earlier expiration or the verifier has a shorter `maxAge`, the earlier takes precedence. The default is `600000`, which is 10 minutes.
* `maxAge`: The maximum allowed age (in seconds) for tokens to still be valid. It is expressed in seconds or a string describing a time span. E.g.: `60`, `"2 days"`, `"10h"`, `"7d"`. A numeric value is interpreted as a seconds count. If you use a string be sure you provide the time units (days, hours, etc.), otherwise milliseconds unit is used by default (`"120"` is equal to `"120ms"`). By default this is not checked.
```

```

> ## `jsonwebtoken` migration changes

Reviewer 2: We should use something like: 'Upgrading from `fastify-jwt@3`'
WDYT ?

Author: Yeah that could be also - I have no strong preference. Maybe others can say something as well?
@simoneb  @Eomm

Reviewer 4: yes let's specify which version we're migrating from. note that this could also live in a standalone doc instead of cluttering the main readme

Author: Done

Migrating from `jsonwebtoken` to `fast-jwt` introduced the following breaking changes:
- **sign** options:
  - `audience` should be changed to `aud`
  - `issuer` should be changed to `iss`
  - `jwtid` should be changed to `jti`
  - `subject` should be changed to `sub`
  - `keyId` should be changed to `kid`

- **verify** options:
  - `audience` should be changed to `allowedAud`
  - `issuer` should be changed to `allowedIss`
  - `subject` should be changed to `allowedSub`

> 

Reviewer 2: ```suggestion
  - `jwtid` should be changed to `allowedJti`
  - `nonce` should be changed to `allowedNonce`

```

Author: Good catch
